### PR TITLE
Fix: Accelerate packaging in ModelBuilder

### PR DESCRIPTION
**Pull request by samruds**
*Issue #, if available:*

Since accelerate is an extra dependencies for Model size calculations fix logic such that we log error if module not present. The flow continues as usual in the absence of accelerate making it an opt in feature for model size calculations and placement. 

Ensure dependency manager for torch servers installs dependencies. 

Also enabled the pt test and re-ran it. 

Happy case logic: 
Module is installed and we pass memory_size_mb which is checked in ModelBuilder._can_fit_on_single_gpu to return True or False

Module not installed
1. Catch import error in _total_inference_model_size_mib
2. Throw ValueError to ModelBuilder._can_fit_on_single_gpu and return False

*Description of changes:*

*Testing done:*

Integ test:

```
py310 run-test: commands[5] | pytest -v -s tests/integ/sagemaker/serve/test_serve_pt_happy.py::test_happy_pytorch_sagemaker_endpoint
sagemaker.config INFO - Not applying SDK defaults from location: /etc/xdg/sagemaker/config.yaml
sagemaker.config INFO - Not applying SDK defaults from location: /home/samruds/.config/sagemaker/config.yaml
======================================================================================================================================================= test session starts =======================================================================================================================================================
platform linux -- Python 3.10.13, pytest-6.2.5, py-1.11.0, pluggy-1.4.0 -- /local/home/samruds/sagemaker-python-sdk/.tox/py310/bin/python
cachedir: .tox/py310/.pytest_cache
rootdir: /local/home/samruds/sagemaker-python-sdk, configfile: tox.ini
plugins: time-machine-2.14.0, anyio-4.3.0, rerunfailures-10.2, forked-1.6.0, cov-3.0.0, xdist-2.4.0, timeout-2.1.0
collected 1 item

tests/integ/sagemaker/serve/test_serve_pt_happy.py::test_happy_pytorch_sagemaker_endpoint[model_builder_inference_spec_schema_builder] ModelBuilder: WARNING:     model_path provided with no image_uri. Attempting to autodetect the image                    by loading the model using inference_spec.load()...
ModelBuilder: INFO:     Autodetecting image since image_uri was not provided in ModelBuilder()
ModelBuilder: WARNING:     Auto detection is only supported for single models DLCs with a framework backend.
ModelBuilder: INFO:     Autodetected framework is pytorch
ModelBuilder: INFO:     Autodetected framework version is 2.0.1
ModelBuilder: INFO:     Auto detected 763104351884.dkr.ecr.us-west-2.amazonaws.com/pytorch-inference:2.0.1-cpu-py310. Proceeding with the the deployment.
ModelBuilder: DEBUG:     Uploading the model resources to bucket=sagemaker-us-west-2-480675430175, key_prefix=session-default-prefix/pytorch-inference-2024-03-29-20-35-59-270.
Uploading model artifacts: 100%|█████████████████| 7901680/7901680 [00:00<00:00, 12174675.61bytes/s]
ModelBuilder: DEBUG:     Model resources uploaded to: s3://sagemaker-us-west-2-480675430175/session-default-prefix/pytorch-inference-2024-03-29-20-35-59-270/serve.tar.gz
ModelBuilder: INFO:     ModelBuilder will collect telemetry to help us better understand our user's needs, diagnose issues, and deliver additional features. To opt out of telemetry, please disable via TelemetryOptOut in intelligent defaults. See https://sagemaker.readthedocs.io/en/stable/overview.html#configuring-and-using-defaults-with-the-sagemaker-python-sdk for more info.
ModelBuilder: DEBUG:     ModelBuilder metrics emitted.
PASSED

======================================================================================================================================================== warnings summary =========================================================================================================================================================
.tox/py310/lib/python3.10/site-packages/invoke/loader.py:3
  /local/home/samruds/sagemaker-python-sdk/.tox/py310/lib/python3.10/site-packages/invoke/loader.py:3: DeprecationWarning: the imp module is deprecated in favour of importlib and slated for removal in Python 3.12; see the module's documentation for alternative uses
    import imp

-- Docs: https://docs.pytest.org/en/stable/warnings.html
============================================================================================================================================ 1 passed, 1 warning in 220.53s (0:03:40) =============================================================================================================================================
_____________________________________________________________________________________________________________________________________________________________ summary _____________________________________________________________________________________________________________________________________________________________
  py310: commands succeeded
  congratulations :)
(pysdk-dev-310) [samruds@dev-dsk-samruds-2c-e20b7667 sagemaker-python-sdk]$
```

Notebook: Happy path

```
from sagemaker.serve.builder.model_builder import ModelBuilder
from sagemaker.serve.builder.schema_builder import SchemaBuilder
from sagemaker.serve import Mode
import json
​
import requests
import json
​
# Run inference
url = "http://localhost:8080/invocations"
​
sample_input = {
 "inputs": "The man worked as a [MASK].",
}
​
headers = {'Content-Type': 'application/json', "Accept":"application/json"}
res = requests.post(url, data=json.dumps(sample_input), headers=headers, timeout=60)
loaded_response = json.loads(res.text)
​
model_builder = ModelBuilder(model="bert-base-uncased",
                             mode=Mode.LOCAL_CONTAINER,
                             schema_builder=SchemaBuilder(sample_input, loaded_response),
                             instance_type='ml.g4dn.xlarge')
sagemaker.config INFO - Not applying SDK defaults from location: /etc/xdg/sagemaker/config.yaml
sagemaker.config INFO - Not applying SDK defaults from location: /home/ec2-user/.config/sagemaker/config.yaml

%%time
builder = model_builder.build()
INFO:botocore.credentials:Found credentials from IAM Role: BaseNotebookInstanceEc2InstanceRole
ModelBuilder: WARNING:     HuggingFace JumpStart Model ID not detected. Building for HuggingFace Model ID.
ModelBuilder: INFO:     GPU Info [g4dn.xlarge]: (1, 16384)
ModelBuilder: INFO:     GPU info (1, 16384) for instance ml.g4dn.xlarge
/home/ec2-user/anaconda3/envs/pytorch_p310/lib/python3.10/site-packages/torch/cuda/__init__.py:628: UserWarning: Can't initialize NVML
  warnings.warn("Can't initialize NVML")
Loading pretrained config for `bert-base-uncased` from `transformers`...
ModelBuilder: INFO:     Total memory size MIB: 501.179296875
ModelBuilder: INFO:     Total memory size MIB: 501.179296875
ModelBuilder: INFO:     Total inference model size MIB 501.179296875, single GPU size for instance MIB 16384.0
ModelBuilder: WARNING:     CUDA is not enabled on your device. Command '['nvidia-smi', '--query-gpu=name,memory.free', '--format=csv']' returned non-zero exit status 9.. Please run ModelBuilder on CUDA enabled hardware to deploy locally.
ModelBuilder: WARNING:     87.62856471008337 percent of disk space used. Please consider freeing up disk space or increasing the EBS volume if you are on a SageMaker Notebook.
ModelBuilder: WARNING:     87.62856471008337 percent of docker disk space at /var/lib/docker is used. Please consider freeing up disk space or increasing the EBS volume if you are on a SageMaker Notebook.
ModelBuilder: INFO:     {'HF_MODEL_ID': 'bert-base-uncased'}
Loading pretrained config for `bert-base-uncased` from `transformers`...
ModelBuilder: INFO:     Detected 763104351884.dkr.ecr.us-east-1.amazonaws.com/huggingface-pytorch-inference:1.7.1-transformers4.6.1-cpu-py36-ubuntu18.04. Proceeding with the the deployment.
CPU times: user 2.25 s, sys: 403 ms, total: 2.66 s
Wall time: 4.79 s
```

Notebook: Force uninstalled accelerate

```
from sagemaker.serve.builder.model_builder import ModelBuilder
from sagemaker.serve.builder.schema_builder import SchemaBuilder
from sagemaker.serve import Mode
import json
​
import requests
import json
​
# Run inference
url = "http://localhost:8080/invocations"
​
sample_input = {
 "inputs": "The man worked as a [MASK].",
}
​
headers = {'Content-Type': 'application/json', "Accept":"application/json"}
res = requests.post(url, data=json.dumps(sample_input), headers=headers, timeout=60)
loaded_response = json.loads(res.text)
​
model_builder = ModelBuilder(model="bert-base-uncased",
                             mode=Mode.LOCAL_CONTAINER,
                             schema_builder=SchemaBuilder(sample_input, loaded_response),
                             instance_type='ml.g4dn.xlarge')
sagemaker.config INFO - Not applying SDK defaults from location: /etc/xdg/sagemaker/config.yaml
sagemaker.config INFO - Not applying SDK defaults from location: /home/ec2-user/.config/sagemaker/config.yaml
%%time
builder = model_builder.build()
INFO:botocore.credentials:Found credentials from IAM Role: BaseNotebookInstanceEc2InstanceRole
ModelBuilder: WARNING:     HuggingFace JumpStart Model ID not detected. Building for HuggingFace Model ID.
ModelBuilder: INFO:     GPU Info [g4dn.xlarge]: (1, 16384)
ModelBuilder: INFO:     GPU info (1, 16384) for instance ml.g4dn.xlarge
ModelBuilder: ERROR:     To enable Model size calculations: Install HuggingFace extras dependencies using pip install 'sagemaker[huggingface]>=2.212.0'
ModelBuilder: INFO:     Unable to determine single GPU size for instance ml.g4dn.xlarge
ModelBuilder: WARNING:     CUDA is not enabled on your device. Command '['nvidia-smi', '--query-gpu=name,memory.free', '--format=csv']' returned non-zero exit status 9.. Please run ModelBuilder on CUDA enabled hardware to deploy locally.
ModelBuilder: WARNING:     88.19689381471613 percent of disk space used. Please consider freeing up disk space or increasing the EBS volume if you are on a SageMaker Notebook.
ModelBuilder: WARNING:     88.19689381471613 percent of docker disk space at /var/lib/docker is used. Please consider freeing up disk space or increasing the EBS volume if you are on a SageMaker Notebook.
ModelBuilder: INFO:     {'HF_MODEL_ID': 'bert-base-uncased'}
ModelBuilder: INFO:     Detected 763104351884.dkr.ecr.us-east-1.amazonaws.com/huggingface-pytorch-inference:1.7.1-transformers4.6.1-cpu-py36-ubuntu18.04. Proceeding with the the deployment.
CPU times: user 776 ms, sys: 33.5 ms, total: 809 ms
Wall time: 1.92 s

```

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [X] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [X] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [X] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [X] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [X] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [X] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [X] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by jiapinw — CHANGES_REQUESTED** [review]

File: src/sagemaker/serve/utils/hardware_detector.py [hunk 1]
```diff
@@ -116,18 +114,27 @@ def _format_instance_type(instance_type: str) -> str:
...
      to add up to an additional 20% to the given model size as found by EleutherAI.
     """
-    args = estimate_command_parser().parse_args([model, "--dtypes", dtype])
-
-    output = gather_data(
-        args
-    )  # "dtype", "Largest Layer", "Total Size Bytes", "Training using Adam"
+    output = None
+    try:
+        from accelerate.commands.estimate import estimate_command_parser, gather_data
```
[the remark this answers is not in the document]
Comment (samruds) [reply src/sagemaker/serve/utils/hardware_detector.py:126]: We still want the feature to be enabled for new PyTorch dist, so it is better for containers to install it. This logic is for old containers which are unable to be re-packaged with accelerate (backward compatible).

[the remark this answers is not in the document]
Comment (jiapinw) [reply src/sagemaker/serve/utils/hardware_detector.py:126]: Makes sense.

**Review by jiapinw — APPROVED** [review]

**Comment by samruds** [comment]
py310 run-test: commands[5] | pytest tests/unit/sagemaker/serve/utils/test_hardware_detector.py
===================================================================== test session starts =====================================================================
platform linux -- Python 3.10.13, pytest-6.2.5, py-1.11.0, pluggy-1.4.0
cachedir: .tox/py310/.pytest_cache
rootdir: /local/home/samruds/sagemaker-python-sdk, configfile: tox.ini
plugins: time-machine-2.14.0, anyio-4.3.0, rerunfailures-10.2, forked-1.6.0, cov-3.0.0, xdist-2.4.0, timeout-2.1.0
collected 8 items

tests/unit/sagemaker/serve/utils/test_hardware_detector.py ........                                                                                     [100%]

====================================================================== warnings summary =======================================================================
.tox/py310/lib/python3.10/site-packages/invoke/loader.py:3
  /local/home/samruds/sagemaker-python-sdk/.tox/py310/lib/python3.10/site-packages/invoke/loader.py:3: DeprecationWarning: the imp module is deprecated in favour of importlib and slated for removal in Python 3.12; see the module's documentation for alternative uses
    import imp

-- Docs: https://docs.pytest.org/en/stable/warnings.html
================================================================ 8 passed, 1 warning in 1.78s =================================================================
___________________________________________________________________________ summary ___________________________________________________________________________
  py310: commands succeeded
  congratulations :)

**Comment by samruds** [comment]
py310 run-test: commands[5] | pytest tests/unit/sagemaker/serve/builder/test_model_builder.py
===================================================================== test session starts =====================================================================
platform linux -- Python 3.10.13, pytest-6.2.5, py-1.11.0, pluggy-1.4.0
cachedir: .tox/py310/.pytest_cache
rootdir: /local/home/samruds/sagemaker-python-sdk, configfile: tox.ini
plugins: time-machine-2.14.0, anyio-4.3.0, rerunfailures-10.2, forked-1.6.0, cov-3.0.0, xdist-2.4.0, timeout-2.1.0
collected 29 items

tests/unit/sagemaker/serve/builder/test_model_builder.py .............................                                                                  [100%]

====================================================================== warnings summary =======================================================================
.tox/py310/lib/python3.10/site-packages/invoke/loader.py:3
  /local/home/samruds/sagemaker-python-sdk/.tox/py310/lib/python3.10/site-packages/invoke/loader.py:3: DeprecationWarning: the imp module is deprecated in favour of importlib and slated for removal in Python 3.12; see the module's documentation for alternative uses
    import imp

-- Docs: https://docs.pytest.org/en/stable/warnings.html
================================================================ 29 passed, 1 warning in 1.60s ================================================================
___________________________________________________________________________ summary ___________________________________________________________________________
  py310: commands succeeded
  congratulations :)

**Comment by samruds** [comment]
```
py310 run-test: commands[5] | pytest tests/unit/sagemaker/serve/detector/test_dependency_manager.py
===================================================================== test session starts =====================================================================
platform linux -- Python 3.10.13, pytest-6.2.5, py-1.11.0, pluggy-1.4.0
cachedir: .tox/py310/.pytest_cache
rootdir: /local/home/samruds/sagemaker-python-sdk, configfile: tox.ini
plugins: time-machine-2.14.0, anyio-4.3.0, rerunfailures-10.2, forked-1.6.0, cov-3.0.0, xdist-2.4.0, timeout-2.1.0
collected 3 items

tests/unit/sagemaker/serve/detector/test_dependency_manager.py ...                                                                                      [100%]

====================================================================== warnings summary =======================================================================
.tox/py310/lib/python3.10/site-packages/invoke/loader.py:3
  /local/home/samruds/sagemaker-python-sdk/.tox/py310/lib/python3.10/site-packages/invoke/loader.py:3: DeprecationWarning: the imp module is deprecated in favour of importlib and slated for removal in Python 3.12; see the module's documentation for alternative uses
    import imp

-- Docs: https://docs.pytest.org/en/stable/warnings.html
================================================================ 3 passed, 1 warning in 0.28s =================================================================
___________________________________________________________________________ summary ___________________________________________________________________________
  py310: commands succeeded
  congratulations :)
```

**Review by akrishna1995 — APPROVED** [review]

**Pull request merged**